File: distilled_simon.py
import random
from time import sleep


def clear():
    print("\n"*80)


def save_score(score):
    with open("dist_high_score", "w") as f:
        f.write(str(score))


def get_random_color():
    colors = ["green", "red", "yellow", "blue"]
    return colors[random.randint(0, 3)]


def game_over(score):
    try:
        with open("dist_high_score") as f:
            high_score = int(f.read())

    except FileNotFoundError:
        save_score(score)
        high_score = score

    if score > high_score:
        save_score(score)
        high_score = score

    print(f"Game Over\n{score = }\n{high_score = }")


def main():
    sequence = []
    score = 0
    run = True

    while run:
        sleep(0.25)
        sequence.append(get_random_color())

        for color in sequence:
            print(color)
            sleep(1)
            clear()

        temp_score = 0
        for color in sequence:
            guess = input("color: ").lower()
            clear()

            if guess == color:
                temp_score += 1

                if temp_score > score:
                    score = temp_score

            else:
                game_over(score)
                run = False
                break


if __name__ == "__main__":
    main()

File: DistilledSimon.java
package SimonGame;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;
import java.io.IOException;

public class DistilledSimon {

	public static void clear() {
		for (int i = 0; i < 80; i++) {
			System.out.println();
		}
	}
	
	public static void saveScore(int score) {
		File file = new File("java_high_score.txt");
		
		try {
			file.createNewFile();
			FileWriter writer = new FileWriter("java_high_score.txt");
			String scoreString = Integer.toString(score);
			
			writer.write(scoreString);
			writer.close();
			
		} catch (IOException e) {
			System.out.println("An error has occurred in saveScore");
			e.printStackTrace();
		}
	}
	
	public static String getRandomColor() {
		String[] colors = {"green", "red", "yellow", "blue"};
		Random random = new Random();
		int randIdx = random.nextInt(colors.length);
		String randColor = colors[randIdx];
		return randColor;
	}
	
	public static void gameOver(int score) {
		int highScore;
		try {
			File file = new File("java_high_score.txt");
			Scanner reader = new Scanner(file);
			highScore = reader.nextInt();
			reader.close();
		} catch (FileNotFoundException e) {
			saveScore(score);
			highScore = score;
		}
		
		if (score > highScore) {
			saveScore(score);
			highScore = score;
		}
		
		System.out.println("Game Over\nScore = " + score + "\nHigh Score = " 
							+ highScore);
	}
	
	public static void printSequence(ArrayList<String> sequence) {
		for (int idx = 0; idx < sequence.size(); idx++) {
			System.out.println(sequence.get(idx));
			
			try {
				Thread.sleep(1000);
			} catch (InterruptedException e) {
				System.out.println("Error in printSequence");
				e.printStackTrace();
			}
			
			clear();
		}
	}
	
	public static int guessSequence(ArrayList<String> sequence, int score, Scanner scanner) {
		int tempScore = 0;
		String guess;
		
		for (int idx = 0; idx < sequence.size(); idx++) {
			System.out.print("color: ");
			guess = scanner.next().toLowerCase();
			clear();
			
			if (guess.equals(sequence.get(idx))) {
				tempScore++;
				
				if (tempScore > score) {
					score = tempScore;
				}
			} else {
				gameOver(score);
				score = -1; // Trying to end loop in main so make score < 0
				break;
			}
		}
		return score;
	}
	
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		ArrayList<String> sequence = new ArrayList<String>();
		int score = 0;
		Boolean run = true;
		
		while (run) {
			try {
				Thread.sleep(250);
			} catch (InterruptedException e) {
				System.out.println("Error in main");
				e.printStackTrace();
			}
			
			sequence.add(getRandomColor());
			printSequence(sequence);
			score = guessSequence(sequence, score, scanner);
			
			if (score < 0) { // score set to -1 when game over
				scanner.close();
				break;
			}
		}
	}
}
